### ISSUE TITLE
Tiller Pod is Crashed when values yaml have issue. 

### ISSUE THREAD

**Issue by hemanthavs** [helm/helm#3290]
Logs are bellow
```
[tiller] 2017/12/21 22:25:30 rendering consume-sample chart using values
panic: reflect: Zero(nil) [recovered]
	panic: reflect: Zero(nil)
 goroutine 313 [running]:
text/template.errRecover(0xc420d53098)
	/usr/local/go/src/text/template/exec.go:146 +0x62
panic(0x1b48260, 0xc4203bfa00)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
reflect.Zero(0x0, 0x0, 0x0, 0x1ba7680, 0xc4203bf9f0)
	/usr/local/go/src/reflect/value.go:2122 +0x134
text/template.(*state).validateType(0xc420d53018, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420d52c28)
	/usr/local/go/src/text/template/exec.go:697 +0x700
text/template.(*state).evalArg(0xc420d53018, 0x1cc8440, 0xc420231020, 0x15, 0x0, 0x0, 0x2ab1ae0, 0xc4208aa000, 0x6, 0x4810a6, ...)
	/usr/local/go/src/text/template/exec.go:746 +0x870
```

Proposal, Better to have helm lint before installing chart. So Error can be handled gracefully instead of Tiller pod crash

**Comment by bacongobbler** [comment]
Thanks @hemanthavs. On one hand I agree that linting the chart may be the right approach, but there may be an edge case or two in which enforcing the linter on all installs could cause certain packages to break so I'm a little skeptical about turning it on in the default case.

Would you mind providing an example chart that can make tiller crash? There's not much in this output to describe what exactly caused it to crash, but an example would be very helpful to determine what caused the panic and how we can handle this case better. Thanks!

**Comment by technosophos** [comment]
To prevent this, we should probably place a `recover` outside of the template rendering code to trap the panic and handle/log it. I'll switch this to a bug. And I would guess this is a good starter, since it will just require finding the template entry point and adding a `recover` that logs and converts to an error.

**Comment by scriptonist** [comment]
Hey Can I try to tackle this issue ?

**Comment by bacongobbler** [comment]
Yes! This would be a fantastic starter issue. :)

**Comment by scriptonist** [comment]
Yup I'm starting to work on this.

**Comment by dmaze** [comment]
[helm-template-crash-0.0.1.tar.gz](https://github.com/kubernetes/helm/files/1790435/helm-template-crash-0.0.1.tar.gz) has a minimal Helm chart that causes Tiller to panic; `recover` in the template rendering code seems like it would in fact fix the problem.

The chart contains a single configmap:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-{{ .Chart.Name }}-placeholder
data:
  release: {{ .Release.Name }}
  releaseButNotFoo: {{ list .Release.Name | without "foo" | join "" }}
  releaseButNotFooCorrect: {{ without (list .Release.Name) "foo" | join "" }}
```

The key error here is that `releaseButNotFoo` calls the spring `without` function with its parameters in the wrong order.  If you unpack the chart and run `helm template .` it panics and you get a Go backtrace, eventually pointing back at `k8s.io/helm/pkg/engine.(*Engine).render` (`k8s.io/helm/pkg/engine/engine.go:255`).  An earlier comment suggested `helm lint`, but that panics too.  If you actually try to `helm install` it, the tiller pod crashes, and `kubectl logs` says:

```
[tiller] 2018/03/07 18:39:37 rendering helm-template-crash chart using values
panic: Cannot find without on type string [recovered]
	panic: Cannot find without on type string

goroutine 245 [running]:
text/template.errRecover(0xc42082ae10)
	/usr/local/go/src/text/template/exec.go:146 +0x62
panic(0x18ba920, 0xc4204055b0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
k8s.io/helm/vendor/github.com/Masterminds/sprig.without(0x18ba920, 0xc420405590, 0xc4204055a0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/vendor/github.com/Masterminds/sprig/list.go:237 +0x3cb
reflect.Value.call(0x195f4a0, 0x1cf4a10, 0x13, 0x1c499d7, 0x4, 0xc420d71710, 0x2, 0x2, 0xc420d73080, 0x1, ...)
	/usr/local/go/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0x195f4a0, 0x1cf4a10, 0x13, 0xc420d71710, 0x2, 0x2, 0x186f140, 0xc420d716f8, 0x97)
	/usr/local/go/src/reflect/value.go:302 +0xa4
text/template.(*state).evalCall(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x195f4a0, 0x1cf4a10, 0x13, 0x297f3e0, 0xc420d71350, 0xc420ddc774, ...)
	/usr/local/go/src/text/template/exec.go:668 +0x51f
text/template.(*state).evalFunction(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420d71380, 0x297f3e0, 0xc420d71350, 0xc420d72f20, 0x2, 0x2, ...)
	/usr/local/go/src/text/template/exec.go:536 +0x18e
text/template.(*state).evalCommand(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420d71350, 0x186f140, 0xc420d716f8, 0x97, 0x186f140, 0xc420d716f8, ...)
	/usr/local/go/src/text/template/exec.go:433 +0x6cd
text/template.(*state).evalPipeline(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420a9d630, 0x0, 0x0, 0x98)
	/usr/local/go/src/text/template/exec.go:406 +0xf2
text/template.(*state).walk(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x297f260, 0xc420d71410)
	/usr/local/go/src/text/template/exec.go:232 +0x55a
text/template.(*state).walk(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x297f5c0, 0xc420d70f60)
	/usr/local/go/src/text/template/exec.go:240 +0x139
text/template.(*Template).execute(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0x1af1f40, 0xc420d70bd0, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:195 +0x20a
text/template.(*Template).Execute(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0x1af1f40, 0xc420d70bd0, 0xc420d71601, 0xc420405420)
	/usr/local/go/src/text/template/exec.go:178 +0x53
text/template.(*Template).ExecuteTemplate(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0xc420d70e40, 0x25, 0x1af1f40, 0xc420d70bd0, 0x1, 0x11d)
	/usr/local/go/src/text/template/exec.go:165 +0x77
k8s.io/helm/pkg/engine.(*Engine).render(0xc42031e320, 0xc420d70de0, 0xc420d70de0, 0x295a9e0, 0xc42031e320)
	/go/src/k8s.io/helm/pkg/engine/engine.go:255 +0xcc6
k8s.io/helm/pkg/engine.(*Engine).Render(0xc42031e320, 0xc420d5ec00, 0xc420d70bd0, 0xc42031e320, 0x1, 0xc42035c000)
	/go/src/k8s.io/helm/pkg/engine/engine.go:123 +0x6a
k8s.io/helm/pkg/tiller.(*ReleaseServer).renderResources(0xc4202dd230, 0xc420d5ec00, 0xc420d70bd0, 0xc420d70b10, 0xc420404e40, 0xc42038f9d3, 0x5, 0x100, 0x1, 0xc420d72d40, ...)
	/go/src/k8s.io/helm/pkg/tiller/release_server.go:266 +0x1aa
k8s.io/helm/pkg/tiller.(*ReleaseServer).prepareRelease(0xc4202dd230, 0xc420d5eba0, 0xc42038fa50, 0x1, 0x1)
	/go/src/k8s.io/helm/pkg/tiller/release_install.go:87 +0x21d
k8s.io/helm/pkg/tiller.(*ReleaseServer).InstallRelease(0xc4202dd230, 0x7f264d855630, 0xc420d0bf50, 0xc420d5eba0, 0xc4202dd230, 0x1c9d07f, 0xe)
	/go/src/k8s.io/helm/pkg/tiller/release_install.go:36 +0xff
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_InstallRelease_Handler.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0x33, 0xc420a9ce10, 0x83b701, 0x2a1fd50)
	/go/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:1228 +0x86
k8s.io/helm/vendor/github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0xc420d5dfa0, 0xc420d5dfc0, 0xc4200ee7f0, 0xc420d5df20, 0xc42077f960, 0x40fe28)
	/go/src/k8s.io/helm/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/server_metrics.go:86 +0xdb
k8s.io/helm/pkg/tiller.newUnaryInterceptor.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0xc420d5dfa0, 0xc420d5dfc0, 0x0, 0xc42077f9d8, 0x40fe28, 0x50)
	/go/src/k8s.io/helm/pkg/tiller/server.go:60 +0xb5
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_InstallRelease_Handler(0x1c03f40, 0xc4202dd230, 0x7f264d855630, 0xc420d0bf50, 0xc420a9cc80, 0x1cf4608, 0x0, 0x0, 0x2, 0x2)
	/go/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:1230 +0x177
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500, 0xc4202dd320, 0x29eb8c8, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:826 +0xc41
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500, 0x0)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:1023 +0x15a6
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4202acb80, 0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:572 +0xa9
created by k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:573 +0xa1
```

```% helm version
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
```

**Comment by scriptonist** [comment]
@dmaze Thank you so much for the detailed description of the problem and Its ***solution***. Since I am very new to helm, I was unable to find a chart to reproduce this issue. Thanks for helping me with this :heart:

**Comment by dmzoneill** [comment]
can somone explain why this alos casues teh above error?

`{{ define "INCLUDE_FILES" }}
{{ (.Files.Glob "resources/*").AsConfig | indent 2 }}
{{ end }}
{{ range $pod := .Values.topology.pods }}
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: vcmts-configmap-vcmtspod
  namespace: default
data:
{{ template "INCLUDE_FILES" }}
{{ end }}
`

**Comment by dmaze** [comment]
At a purely mechanical level, this line is wrong:

```
{{ template "INCLUDE_FILES" }}
```

You probably need to pass the current context to it, more like

```
{{ template "INCLUDE_FILES" . }}
```

When the included template is invoked, `.` will be Go `nil`, and so `.Files` will be an error.  It shouldn't panic, though.

**Comment by bacongobbler** [comment]
I believe this is fixed with Go 1.12 choosing to change the behaviour of a `panic` when rendering a text/template. #5382 fixes this for Helm 3, but it'd be an easy patch for someone to port for Helm 2.

**Comment by bacongobbler** [comment]
ref: https://golang.org/doc/go1.12#text/template

**Comment by hayorov** [comment]
> I believe this is fixed with Go 1.12 choosing to change the behaviour of a `panic` when rendering a text/template. #5382 fixes this for Helm 3, but it'd be an easy patch for someone to port for Helm 2.

@bacongobbler looks like this issue could be closed, go version bumped in https://github.com/helm/helm/commit/470203a173942dac23396dbba575cc9b16253486

**Comment by Alialkhathami** [comment]
Good
